src/classlibnative/bcltype/arraynative.inl

```diff
@@ -0,0 +1,248 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+//
+// File: ArrayNative.cpp
+//
+
+//
+// This file contains the native methods that support the Array class
+//
+
+#ifndef _ARRAYNATIVE_INL_
+#define _ARRAYNATIVE_INL_
+
+#include "gchelpers.inl"
+
+FORCEINLINE void InlinedForwardGCSafeCopyHelper(void *dest, const void *src, size_t len)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_COOPERATIVE;
+        PRECONDITION(CheckPointer(dest));
+        PRECONDITION(CheckPointer(src));
+        SO_TOLERANT;
+    }
+    CONTRACTL_END;
+
+    _ASSERTE(dest != nullptr);
+    _ASSERTE(src != nullptr);
+    _ASSERTE(dest != src);
+    _ASSERTE(len != 0);
+
+    // To be able to copy forwards, the destination buffer cannot start inside the source buffer
+    _ASSERTE((SIZE_T)dest - (SIZE_T)src >= len);
+
+    // Make sure everything is pointer aligned
+    _ASSERTE(IS_ALIGNED(dest, sizeof(SIZE_T)));
+    _ASSERTE(IS_ALIGNED(src, sizeof(SIZE_T)));
+    _ASSERTE(IS_ALIGNED(len, sizeof(SIZE_T)));
+
+    _ASSERTE(CheckPointer(dest));
+    _ASSERTE(CheckPointer(src));
+
+    SIZE_T *dptr = (SIZE_T *)dest;
+    SIZE_T *sptr = (SIZE_T *)src;
+
+    if ((len & sizeof(SIZE_T)) != 0)
+    {
+        *dptr = *sptr;
+        len ^= sizeof(SIZE_T);
+        if (len == 0)
+        {
+            return;
+        }
+        ++sptr;
+        ++dptr;
+    }
+
+    if ((len & (2 * sizeof(SIZE_T))) != 0)
+    {
+#if defined(_AMD64_) && !defined(FEATURE_PAL) // TODO: Enable on Unix
+        __m128 v = _mm_loadu_ps((float *)sptr);
+        _mm_storeu_ps((float *)dptr, v);
+#else // !_AMD64_ || FEATURE_PAL
+        // Read two values and write two values to hint the use of wide loads and stores
+        SIZE_T p[2];
```

Comment:
It is hard to tell what the different compilers are going to do with this. I think it would be better to keep the original implementation as the portable default; and keep then have platform or compiler specific implementations under one big ifdefs - there will be some duplication between the different big block ifdefs, but I think it is fine.